base/reduce.jl

```diff
@@ -954,30 +953,30 @@ julia> count([true, false, true, true])
 3
 ```
 """
-count(itr) = count(identity, itr)
+count(itr; init=0) = count(identity, itr; init)
 
-count(f, itr) = _simple_count(f, itr)
+count(f, itr; init=0) = _simple_count(f, itr, init)
 
-function _simple_count(pred, itr)
-    n = 0
+function _simple_count(pred, itr, init)
+    n = init
     for x in itr
         n += pred(x)::Bool
     end
     return n
 end
 
-function count(::typeof(identity), x::Array{Bool})
-    n = 0
+function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
```

Comment:
I think we need to restrict `T` to `Integer` if we keep the return value type guarantee in the docstring signature (`-> Integer`):

```suggestion
function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T <: Integer}
```

Reply:
(Removing `-> Integer` from the docstring and calling this PR a minor change is, technically, another option.)

Reply:
It feels a bit arbitrary to restrict `T` to only `Integer` types, since there's really nothing preventing this from working with any other accumulation types. The docstring wasn't even that accurate in the first place, since `count` always returned `Int`, so I don't know why it's documented as `Integer`. I always thought "minor change" meant it was slightly breaking for some edge cases, which I don't think applies here, since the previous behavior should be completely maintained, just extended. But you are right, the docstring still needs to include the `dims` argument and have a compat annotation.

Reply:
> there's really nothing preventing this from working with any other accumulation types

If you need other accumulation types, I think you should use `sum`.

> The docstring wasn't even that accurate in the first place, since `count` always returned `Int`

`Int <: Integer` so the docstring in Julia <= 1.5 is OK in the sense `c = count(...); c::Integer` never throws if `count` returns.

But I do agree that docstring is not accurate in the sense that what `->` means is not defined anywhere (i.e., we don't have a specification of the signature part of the docstring).

> I always thought "minor change" meant it was slightly breaking for some edge cases

"Minor change" means to violate [SemVer](https://semver.org) guarantee (= "technically breaking"). See also https://julialang.org/blog/2019/08/release-process/#minor_releases

> since the previous behavior should be completely maintained

Breaking or not is decided with respect to the public API _specification_ (in `julia`, it's documentation). We then check if the behavior follows the API; if not, it's a bug.

So, increasing the return value type bound (`Integer` to `Any`) in the documentation is breaking.

Reply:
Since it won't break any code in practice, I don't think the breakage here would be any kind of problem, but I'm ok with restricting it to `Integer` as well. I guess that restriction should be in the definition of `count` though, not the internal `_simple_count`.

Reply:
IIUC, when not specifying `init`, the result type is `Integer`, in line with the old docs. I.e. old signatures:
```
  count(p, itr) -> Integer
  count(itr) -> Integer
```
New signatures:
```
  count(p, itr) -> Integer
  count(itr) -> Integer
  count(p, itr; init::T) -> T
  count(itr; init::T) -> T
```
Where the latter can be condensed to the equivalent
```
  count(p, itr; init::T=0) -> T
  count(itr; init::T=0) -> T
```
So this seems to be purely an extension; I don't see anything technically breaking here. (For simplicity, this assumes `+(::T,::T)::T`, but that holds for `T==Int`.)

Reply:
So should I leave this as is, or does it make more sense to constrain this to `Integer`?

Reply:
@tkf I'm fine with either option, should I restrict `init` to `Integer or not?

Reply:
I don't have a strong opinion here, but I think not forcing `init isa Integer` is fine.